### ISSUE TITLE
[BUG][Azure.Identity] ManagedIdentitySource failed to parse response

### ISSUE THREAD

**Issue by zhiweiv** [Azure/azure-sdk-for-net#24158]
**Describe the bug**
```
Azure.Identity.AuthenticationFailedException: ManagedIdentityCredential authentication failed: 'p' is an invalid start of a value. LineNumber: 0 | BytePositionInLine: 0.
 ---> System.Text.Json.JsonReaderException: 'p' is an invalid start of a value. LineNumber: 0 | BytePositionInLine: 0.
   at System.Text.Json.JsonDocument.Parse(Stream utf8Json, JsonDocumentOptions options)
   at Azure.Identity.ManagedIdentitySource.HandleResponseAsync(Boolean async, TokenRequestContext context, Response response, CancellationToken cancellationToken)
   at Azure.Identity.ImdsManagedIdentitySource.HandleResponseAsync(Boolean async, TokenRequestContext context, Response response, CancellationToken cancellationToken)
```

Here is the source code, seems the IMDS doesn't always return json response, this method converts json before responde code check.
```
        protected virtual async ValueTask<AccessToken> HandleResponseAsync(bool async, TokenRequestContext context, Response response, CancellationToken cancellationToken)
        {
            using JsonDocument json = async
                ? await JsonDocument.ParseAsync(response.ContentStream, default, cancellationToken).ConfigureAwait(false)
                : JsonDocument.Parse(response.ContentStream);
            if (response.Status == 200)
            {
                return GetTokenFromResponse(json.RootElement);
            }

            string message = GetMessageFromResponse(json.RootElement);
            throw async
                ? await Pipeline.Diagnostics.CreateRequestFailedExceptionAsync(response, message).ConfigureAwait(false)
                : Pipeline.Diagnostics.CreateRequestFailedException(response, message);
        }
```

**Environment:**
 - Name and version of the Library package used: Azure.Identity 1.5.0-beta.4
 - Hosting platform or OS and .NET runtime version: .NET 5.0 in AKS Linux container

**Comment by jsquire** [comment]
Thank you for your feedback.  Tagging and routing to the team members best able to assist.

**Comment by zhiweiv** [comment]
A similiar error
```
Azure.Identity.AuthenticationFailedException: ManagedIdentityCredential authentication failed: 'failed to refresh token, error: the MSI endpoint is not available. Failed HTTP request to MSI endpoint: Get "http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01": context deadline exceeded
' is an invalid JSON literal. Expected the literal 'false'. LineNumber: 0 | BytePositionInLine: 2.
 ---> System.Text.Json.JsonReaderException: 'failed to refresh token, error: the MSI endpoint is not available. Failed HTTP request to MSI endpoint: Get "http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01": context deadline exceeded
' is an invalid JSON literal. Expected the literal 'false'. LineNumber: 0 | BytePositionInLine: 2.
   at System.Text.Json.JsonDocument.Parse(Stream utf8Json, JsonDocumentOptions options)
   at Azure.Identity.ManagedIdentitySource.HandleResponseAsync(Boolean async, TokenRequestContext context, Response response, CancellationToken cancellationToken)
   at Azure.Identity.ImdsManagedIdentitySource.HandleResponseAsync(Boolean async, TokenRequestContext context, Response response, CancellationToken cancellationToken)
```

**Comment by zhiweiv** [comment]
@christothes @schaabs 
Do you know when will have a new release(or beta release) of Azure.Identity which contains the fix? We encounter more similar error, we'd like to get the raw exception message to find the root cause.

**Comment by christothes** [comment]
Probably not until end of January.

**Comment by fonbrauzen** [comment]
@christothes any news after the end of January?

**Comment by christothes** [comment]
Hi @fonbrauzen - We should be releasing next week (It is usually the 2nd Tuesday of the month).

**Comment by zhiweiv** [comment]
Any update?

**Comment by zhiweiv** [comment]
1.6.0-beta.1 released.

**Comment by mamort** [comment]
Can you please reopen this @zhiweiv ? The fix #24166 does not seem to properly fix this issue. If the IMDS returns a http error like 500, 401 then the response probably does not contain valid json in the body? At least we have an issue where we just get:
```
Azure.Identity.CredentialUnavailableException: Managed Identity response was not in the expected format. See the inner exception for details.
 ---> System.Text.Json.JsonReaderException: 'a' is an invalid start of a value. LineNumber: 0 | BytePositionInLine: 0.
   at System.Text.Json.ThrowHelper.ThrowJsonReaderException(Utf8JsonReader& json, ExceptionResource resource, Byte nextByte, ReadOnlySpan`1 bytes)
   at System.Text.Json.Utf8JsonReader.ConsumeValue(Byte marker)
   at System.Text.Json.Utf8JsonReader.ReadFirstToken(Byte first)
   at System.Text.Json.Utf8JsonReader.ReadSingleSegment()
   at System.Text.Json.Utf8JsonReader.Read()
   at System.Text.Json.JsonDocument.Parse(ReadOnlySpan`1 utf8JsonSpan, JsonReaderOptions readerOptions, MetadataDb& database, StackRowStack& stack)
   at System.Text.Json.JsonDocument.Parse(ReadOnlyMemory`1 utf8Json, JsonReaderOptions readerOptions, Byte[] extraRentedArrayPoolBytes, PooledByteBufferWriter extraPooledByteBufferWriter)
   at System.Text.Json.JsonDocument.ParseAsyncCore(Stream utf8Json, JsonDocumentOptions options, CancellationToken cancellationToken)
   at Azure.Identity.ManagedIdentitySource.HandleResponseAsync(Boolean async, TokenRequestContext context, Response response, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at Azure.Identity.ManagedIdentitySource.HandleResponseAsync(Boolean async, TokenRequestContext context, Response response, CancellationToken cancellationToken)
   at Azure.Identity.ImdsManagedIdentitySource.HandleResponseAsync(Boolean async, TokenRequestContext context, Response response, CancellationToken cancellationToken)
   at Azure.Identity.ManagedIdentitySource.AuthenticateAsync(Boolean async, TokenRequestContext context, CancellationToken cancellationToken)
   at Azure.Identity.ImdsManagedIdentitySource.AuthenticateAsync(Boolean async, TokenRequestContext context, CancellationToken cancellationToken)
   at Azure.Identity.ManagedIdentityClient.AuthenticateCoreAsync(Boolean async, TokenRequestContext context, CancellationToken cancellationToken)
   (...)
   ```
   
   It would be really useful to at least get the statuscode

**Comment by zhiweiv** [comment]
In our environment, after upgraded to Azure.Identity 1.6.0+, in some cases indeed it failed to parse json, but we can see the status code, see the screenshot we got before.
<img width="802" alt="exception" src="https://user-images.githubusercontent.com/22001089/216212981-5eb3cdac-0585-43d5-a9a3-0e14c2a3cfa0.png">

**Comment by mamort** [comment]
Your stacktrace is a bit different than mine with a `AuthenticationFailedException`. Here is my full trace:
```
Azure.Identity.CredentialUnavailableException: Managed Identity response was not in the expected format. See the inner exception for details.
 ---> System.Text.Json.JsonReaderException: 'a' is an invalid start of a value. LineNumber: 0 | BytePositionInLine: 0.
   at System.Text.Json.ThrowHelper.ThrowJsonReaderException(Utf8JsonReader& json, ExceptionResource resource, Byte nextByte, ReadOnlySpan`1 bytes)
   at System.Text.Json.Utf8JsonReader.ConsumeValue(Byte marker)
   at System.Text.Json.Utf8JsonReader.ReadFirstToken(Byte first)
   at System.Text.Json.Utf8JsonReader.ReadSingleSegment()
   at System.Text.Json.Utf8JsonReader.Read()
   at System.Text.Json.JsonDocument.Parse(ReadOnlySpan`1 utf8JsonSpan, JsonReaderOptions readerOptions, MetadataDb& database, StackRowStack& stack)
   at System.Text.Json.JsonDocument.Parse(ReadOnlyMemory`1 utf8Json, JsonReaderOptions readerOptions, Byte[] extraRentedArrayPoolBytes, PooledByteBufferWriter extraPooledByteBufferWriter)
   at System.Text.Json.JsonDocument.ParseAsyncCore(Stream utf8Json, JsonDocumentOptions options, CancellationToken cancellationToken)
   at Azure.Identity.ManagedIdentitySource.HandleResponseAsync(Boolean async, TokenRequestContext context, Response response, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at Azure.Identity.ManagedIdentitySource.HandleResponseAsync(Boolean async, TokenRequestContext context, Response response, CancellationToken cancellationToken)
   at Azure.Identity.ImdsManagedIdentitySource.HandleResponseAsync(Boolean async, TokenRequestContext context, Response response, CancellationToken cancellationToken)
   at Azure.Identity.ManagedIdentitySource.AuthenticateAsync(Boolean async, TokenRequestContext context, CancellationToken cancellationToken)
   at Azure.Identity.ImdsManagedIdentitySource.AuthenticateAsync(Boolean async, TokenRequestContext context, CancellationToken cancellationToken)
   at Azure.Identity.ManagedIdentityClient.AuthenticateCoreAsync(Boolean async, TokenRequestContext context, CancellationToken cancellationToken)
   at Azure.Identity.ManagedIdentityClient.AppTokenProviderImpl(AppTokenProviderParameters parameters)
   at Microsoft.Identity.Client.Internal.Requests.ClientCredentialRequest.SendTokenRequestToProviderAsync(CancellationToken cancellationToken)
   at Microsoft.Identity.Client.Internal.Requests.ClientCredentialRequest.FetchNewAccessTokenAsync(CancellationToken cancellationToken)
   at Microsoft.Identity.Client.Internal.Requests.ClientCredentialRequest.ExecuteAsync(CancellationToken cancellationToken)
   at Microsoft.Identity.Client.Internal.Requests.RequestBase.RunAsync(CancellationToken cancellationToken)
   at Microsoft.Identity.Client.ApiConfig.Executors.ConfidentialClientExecutor.ExecuteAsync(AcquireTokenCommonParameters commonParameters, AcquireTokenForClientParameters clientParameters, CancellationToken cancellationToken)
   at Azure.Identity.AbstractAcquireTokenParameterBuilderExtensions.ExecuteAsync[T](AbstractAcquireTokenParameterBuilder`1 builder, Boolean async, CancellationToken cancellationToken)
   at Azure.Identity.MsalConfidentialClient.AcquireTokenForClientCoreAsync(String[] scopes, String tenantId, Boolean async, CancellationToken cancellationToken)
   at Azure.Identity.MsalConfidentialClient.AcquireTokenForClientAsync(String[] scopes, String tenantId, Boolean async, CancellationToken cancellationToken)
   at Azure.Identity.ManagedIdentityClient.AuthenticateAsync(Boolean async, TokenRequestContext context, CancellationToken cancellationToken)
   at Azure.Identity.ManagedIdentityCredential.GetTokenImplAsync(Boolean async, TokenRequestContext requestContext, CancellationToken cancellationToken)
   at Azure.Identity.CredentialDiagnosticScope.FailWrapAndThrow(Exception ex, String additionalMessage)
   at Azure.Identity.ManagedIdentityCredential.GetTokenImplAsync(Boolean async, TokenRequestContext requestContext, CancellationToken cancellationToken)
   at Azure.Identity.ManagedIdentityCredential.GetTokenAsync(TokenRequestContext requestContext, CancellationToken cancellationToken)
   at Azure.Core.Pipeline.BearerTokenAuthenticationPolicy.AccessTokenCache.GetHeaderValueFromCredentialAsync(TokenRequestContext context, Boolean async, CancellationToken cancellationToken)
   at Azure.Core.Pipeline.BearerTokenAuthenticationPolicy.AccessTokenCache.GetHeaderValueAsync(HttpMessage message, TokenRequestContext context, Boolean async)
   at Azure.Core.Pipeline.BearerTokenAuthenticationPolicy.AccessTokenCache.GetHeaderValueAsync(HttpMessage message, TokenRequestContext context, Boolean async)
   at Azure.Core.Pipeline.BearerTokenAuthenticationPolicy.AuthenticateAndAuthorizeRequestAsync(HttpMessage message, TokenRequestContext context)
   at Azure.Security.KeyVault.ChallengeBasedAuthenticationPolicy.AuthorizeRequestOnChallengeAsyncInternal(HttpMessage message, Boolean async)
   at Azure.Core.Pipeline.BearerTokenAuthenticationPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)
   at Azure.Core.Pipeline.RedirectPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)
   at Azure.Core.Pipeline.RetryPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)
   at Azure.Core.Pipeline.RetryPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)
   at Azure.Core.Pipeline.HttpPipeline.SendRequestAsync(Request request, CancellationToken cancellationToken)
   at Azure.Security.KeyVault.KeyVaultPipeline.SendRequestAsync(Request request, CancellationToken cancellationToken)
   at Azure.Security.KeyVault.KeyVaultPipeline.GetPageAsync[T](Uri firstPageUri, String nextLink, Func`1 itemFactory, String operationName, CancellationToken cancellationToken)
   at Azure.Core.PageResponseEnumerator.FuncAsyncPageable`1.AsPages(String continuationToken, Nullable`1 pageSizeHint)+MoveNext()
   at Azure.Core.PageResponseEnumerator.FuncAsyncPageable`1.AsPages(String continuationToken, Nullable`1 pageSizeHint)+System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult()
   at Azure.AsyncPageable`1.GetAsyncEnumerator(CancellationToken cancellationToken)+MoveNext()
   at Azure.AsyncPageable`1.GetAsyncEnumerator(CancellationToken cancellationToken)+MoveNext()
   at Azure.AsyncPageable`1.GetAsyncEnumerator(CancellationToken cancellationToken)+System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult()
   at Azure.Extensions.AspNetCore.Configuration.Secrets.AzureKeyVaultConfigurationProvider.LoadAsync()
   at Azure.Extensions.AspNetCore.Configuration.Secrets.AzureKeyVaultConfigurationProvider.LoadAsync()
   at Azure.Extensions.AspNetCore.Configuration.Secrets.AzureKeyVaultConfigurationProvider.Load()
   at Microsoft.Extensions.Configuration.ConfigurationRoot..ctor(IList`1 providers)
   at Microsoft.Extensions.Configuration.ConfigurationBuilder.Build()
   at WebCreditApi.Program.<>c.<CreateHostBuilder>b__1_3(WebHostBuilderContext context, IConfigurationBuilder builder) in /src/src/WebCreditApi/Program.cs:line 93
   at Microsoft.AspNetCore.Hosting.GenericWebHostBuilder.<>c__DisplayClass9_0.<ConfigureAppConfiguration>b__0(HostBuilderContext context, IConfigurationBuilder builder)
   at Microsoft.Extensions.Hosting.HostBuilder.BuildAppConfiguration()
   at Microsoft.Extensions.Hosting.HostBuilder.Build()
   at WebCreditApi.Program.Main(String[] args) in /src/src/WebCreditApi/Program.cs:line 29
   ```
I don't get the statuscode. 

From looking at the sourcecode it seems this is handled in `ManagedIdentitySource.cs` and I can't see how the status code would be logged:
```

        protected virtual async ValueTask<AccessToken> HandleResponseAsync(
            bool async,
            TokenRequestContext context,
            Response response,
            CancellationToken cancellationToken)
        {
            string message;
            Exception exception = null;
            try
            {
                using JsonDocument json = async
                    ? await JsonDocument.ParseAsync(response.ContentStream, default, cancellationToken).ConfigureAwait(false)
                    : JsonDocument.Parse(response.ContentStream);
                if (response.Status == 200)
                {
                    return GetTokenFromResponse(json.RootElement);
                }

                message = GetMessageFromResponse(json.RootElement);
            }
            catch (JsonException jex)
            {
                throw new CredentialUnavailableException(UnexpectedResponse, jex);
            }
            catch (Exception e)
            {
                exception = e;
                message = UnexpectedResponse;
            }

            var responseError = new ResponseError(null, message);
            throw async
                ? await Pipeline.Diagnostics.CreateRequestFailedExceptionAsync(response, responseError, innerException: exception).ConfigureAwait(false)
                : Pipeline.Diagnostics.CreateRequestFailedException(response, responseError, innerException: exception);
        }
```

**Comment by zhiweiv** [comment]
OK, seems my exception was thrown from this part,
```
            catch (Exception e)
            {
                exception = e;
                message = UnexpectedResponse;
            }

            var responseError = new ResponseError(null, message);
            throw async
                ? await Pipeline.Diagnostics.CreateRequestFailedExceptionAsync(response, responseError, innerException: exception).ConfigureAwait(false)
                : Pipeline.Diagnostics.CreateRequestFailedException(response, responseError, innerException: exception);
```

I suggest you open a new issue to trace, this one is too old, most probably not get noticed from MS side.

**Comment by christothes** [comment]
The exception: `Azure.Identity.CredentialUnavailableException: Managed Identity response was not in the expected format` is the expected behavior. It should result in the Managed Identity credential being skipped in the `DefaultAzureCredential` credential chain.

**Comment by devlie** [comment]
> The exception: `Azure.Identity.CredentialUnavailableException: Managed Identity response was not in the expected format` is the expected behavior. It should result in the Managed Identity credential being skipped in the `DefaultAzureCredential` credential chain.

@christothes This must be a bug, IMO (or at least https://github.com/Azure/azure-sdk-for-net/issues/28568 must be). 

IMDS can return HTTP 410 by design and their documentation requires caller to retry up to 70 secs, but `ManagedIdentityCredential` does not retry even once. The chain shouldn't have skipped past `ManagedIdentityCredential` without retrying HTTP 410. In our case, we're using Managed Identity to connect to our resources like AzStorage and CosmosDB, so skipping past MI does no good for us. Worse, because we don't retrieve the token ourselves (Storage/Cosmos clients do that), we can't really retry the error ourselves without polluting our business layer which has no business of knowing which auth is used.
